// @mariozechner/pi-coding-agent/dist/index.js
// Test stub for @mariozechner/pi-coding-agent
// Minimal CustomEditor implementation sufficient for ModalEditor tests.

export class CustomEditor {
  constructor(_tui, _theme, _kb) {
    this.state = { lines: [""], cursorLine: 0, cursorCol: 0 };
    this.preferredVisualCol = 0;
    this.tui = { requestRender() {} };
  }

  handleInput(data) {
    const s = this.state;

    // ESC_RIGHT
    if (data === "\x1b[C") {
      const line = s.lines[s.cursorLine] ?? "";
      if (s.cursorCol < line.length) s.cursorCol++;
      return;
    }
    // ESC_LEFT
    if (data === "\x1b[D") {
      if (s.cursorCol > 0) s.cursorCol--;
      return;
    }
    // ESC_UP
    if (data === "\x1b[A") {
      if (s.cursorLine > 0) {
        s.cursorLine--;
        const line = s.lines[s.cursorLine] ?? "";
        s.cursorCol = Math.min(s.cursorCol, line.length);
      }
      return;
    }
    // ESC_DOWN
    if (data === "\x1b[B") {
      if (s.cursorLine < s.lines.length - 1) {
        s.cursorLine++;
        const line = s.lines[s.cursorLine] ?? "";
        s.cursorCol = Math.min(s.cursorCol, line.length);
      }
      return;
    }
    // ESC_DELETE — delete char at cursor (or join with next line)
    if (data === "\x1b[3~") {
      const line = s.lines[s.cursorLine] ?? "";
      if (s.cursorCol < line.length) {
        s.lines[s.cursorLine] = line.slice(0, s.cursorCol) + line.slice(s.cursorCol + 1);
      } else if (s.cursorLine < s.lines.length - 1) {
        const next = s.lines[s.cursorLine + 1] ?? "";
        s.lines[s.cursorLine] = line + next;
        s.lines.splice(s.cursorLine + 1, 1);
      }
      return;
    }
    // CTRL_A — move to col 0
    if (data === "\x01") {
      s.cursorCol = 0;
      return;
    }
    // CTRL_E — move to end of line
    if (data === "\x05") {
      const line = s.lines[s.cursorLine] ?? "";
      s.cursorCol = line.length;
      return;
    }
    // CTRL_K — kill to end of line (or join with next line)
    if (data === "\x0b") {
      const line = s.lines[s.cursorLine] ?? "";
      if (s.cursorCol < line.length) {
        s.lines[s.cursorLine] = line.slice(0, s.cursorCol);
      } else if (s.cursorLine < s.lines.length - 1) {
        const next = s.lines[s.cursorLine + 1] ?? "";
        s.lines[s.cursorLine] = line + next;
        s.lines.splice(s.cursorLine + 1, 1);
      }
      return;
    }
    // CTRL_UNDERSCORE — undo (no-op in tests)
    if (data === "\x1f") return;
    // NEWLINE — split line at cursor
    if (data === "\n") {
      const line = s.lines[s.cursorLine] ?? "";
      const left = line.slice(0, s.cursorCol);
      const right = line.slice(s.cursorCol);
      s.lines[s.cursorLine] = left;
      s.lines.splice(s.cursorLine + 1, 0, right);
      s.cursorLine++;
      s.cursorCol = 0;
      return;
    }
    // ESC — no-op (passed upward by ModalEditor for agent abort; stub ignores it)
    if (data === "\x1b") return;
    // Tab — insert literally
    if (data === "\t") {
      const line = s.lines[s.cursorLine] ?? "";
      s.lines[s.cursorLine] = line.slice(0, s.cursorCol) + "\t" + line.slice(s.cursorCol);
      s.cursorCol++;
      return;
    }
    // Other control sequences — no-op
    if (data.length > 0 && data.codePointAt(0) < 32) return;

    // Printable single codepoint — insert at cursor
    const codePoints = [...data];
    if (codePoints.length === 1) {
      const cp = data.codePointAt(0);
      if (cp >= 32 && cp !== 127) {
        const line = s.lines[s.cursorLine] ?? "";
        s.lines[s.cursorLine] = line.slice(0, s.cursorCol) + data + line.slice(s.cursorCol);
        s.cursorCol++;
      }
    }
  }

  getLines() { return this.state.lines; }
  getCursor() { return { line: this.state.cursorLine, col: this.state.cursorCol }; }
  render(_width) { return this.state.lines.slice(); }
}

export function copyToClipboard(_text) {}
